### PR TITLE
allArgConstructor for java

### PR DESCRIPTION
**Pull request by jpfinne**
All Argument constructor for java client and spring

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming 7.5.0 minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [ ] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

File: .github/workflows/openapi-generator.yaml [hunk 1]
```diff
@@ -140,8 +140,23 @@ jobs:
...
+          rm -rf samples/client/petstore/csharp/generichost/net8/Petstore
+          rm -rf samples/client/petstore/csharp/generichost/net8/SourceGeneration
+          rm -rf samples/client/petstore/csharp/generichost/net8/UseDateTimeForDate
+          rm -rf samples/client/petstore/csharp/generichost/standard2.0/Petstore
       - name: Generate samples
-        run: bash bin/generate-samples.sh
+        run: |
+          bash bin/generate-samples.sh
+          # when a sample is deleted, you have to generate it twice for all files to get created
+          bash bin/generate-samples.sh
```
Comment (wing328) [inline .github/workflows/openapi-generator.yaml:159]: this PR includes lots of changes not belonging to your original PR

can you please submit a new one with just the change for just "allArgConstruct for java"?

Comment (jpfinne) [reply .github/workflows/openapi-generator.yaml]: sorry. I screwed up the merges.

Is it better now?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java [hunk 2]
```diff
@@ -703,9 +725,101 @@ public Map<String, ModelsMap> postProcessAllModels(Map<String, ModelsMap> objs)
...
+            for (CodegenModel cm : allModels.values()) {
+                if (isConstructorWithAllArgsAllowed(cm)) {
+                    cm.vendorExtensions.put("generatedConstructorWithAllArgs", true);
+                    List<Object> constructorArgs = new ArrayList<>();
+                    // vendorExtensions.allArgsConstructorVars should be equivalent to allVars
+                    // but it is not reliable if openapiNormalizer.REFACTOR_ALLOF_WITH_PROPERTIES_ONLY is disabled
+                    if (cm.vars.size() + cm.parentVars.size() != cm.allVars.size()) {
+                        once(LOGGER).warn("Unexpected allVars for {} expecting:{} vars. actual:{} vars", cm.name, cm.vars.size() + cm.parentVars.size(), cm.allVars.size());
+                    }
+                    cm.vendorExtensions.put("allArgsConstructorVars", constructorArgs);
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java:791]: for extensions, please add `x-` as the prefix and use dash case

e.g. x-java-all-args-constructor-vars

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java [hunk 3]
```diff
@@ -703,9 +725,101 @@ public Map<String, ModelsMap> postProcessAllModels(Map<String, ModelsMap> objs)
...
+        if (isGeneratedConstructorWithAllArgs()) {
+            // conditionally force the generation of all args constructor.
+            for (CodegenModel cm : allModels.values()) {
+                if (isConstructorWithAllArgsAllowed(cm)) {
+                    cm.vendorExtensions.put("generatedConstructorWithAllArgs", true);
+                    List<Object> constructorArgs = new ArrayList<>();
+                    // vendorExtensions.allArgsConstructorVars should be equivalent to allVars
+                    // but it is not reliable if openapiNormalizer.REFACTOR_ALLOF_WITH_PROPERTIES_ONLY is disabled
+                    if (cm.vars.size() + cm.parentVars.size() != cm.allVars.size()) {
+                        once(LOGGER).warn("Unexpected allVars for {} expecting:{} vars. actual:{} vars", cm.name, cm.vars.size() + cm.parentVars.size(), cm.allVars.size());
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java:789]: not sure if we need this warning as vars+ parentVars may not be equal to allVars when vars and parentVars have overlayed properties (e.g. both child and parent model has a property named `id` defined)

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java (hunk 2 above)
Comment (jpfinne) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java]: done

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java (hunk 3 above)
Comment (wing328) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java:790]: >  if (cm.vars.size() + cm.parentVars.size() != cm.allVars.size()) {

let's step back for a moment. did you encounter such condition before? if yes, can you please share the spec?

Comment (jpfinne) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java:790]: You can find more evidences in #18340

So the merge of parentVars and allVars is needed to ensure that the constructors are correctly generated

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/SpringCodegen.java [hunk 4]
```diff
@@ -1168,66 +1168,19 @@ public CodegenModel fromModel(String name, Schema model) {
...
-                // There must be a better way ...
-                for (String imp: inheritedImports) {
-                    String qimp = importMapping().get(imp);
-                    if (qimp != null) {
-                        Map<String,String> toAdd = new HashMap<>();
-                        toAdd.put("import", qimp);
-                        modelsAttrs.getImports().add(toAdd);
-                    }
-                }
-            }
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/SpringCodegen.java:1225]: fyi above code block refactored into abstract java codegen.

@cachescrubber (2022/02) @welshm (2022/02) @MelleD (2022/02) @atextor (2022/02) @manedev79 (2022/02) @javisst (2022/02) @borsch (2022/02) @banlevente (2022/02) @Zomzog (2022/09) @martin-mfg (2023/08)

**Comment by wing328** [comment]
fyi. i've filed https://github.com/OpenAPITools/openapi-generator/pull/18538 with resolved merge conflicts and minor change (removing EOL in pojo)

**Comment by wing328** [comment]
cc 
@bbdouglas (2017/07) @sreeshas (2017/08) @jfiala (2017/08) @lukoyanov (2017/09) @cbornet (2017/09) @jeff9finger (2018/01) @karismann (2019/03) @Zomzog (2019/04) @lwlee2608 (2019/10) @martin-mfg (2023/08)

**Comment by wing328** [comment]
closed via https://github.com/OpenAPITools/openapi-generator/pull/18538

**Pull request closed without merging**